### PR TITLE
refactor(GlobalHeader): DRY up HeaderItems

### PR DESCRIPTION
**Pull request by j10sanders**
### Overview

Just a small DRYing up of two functions that have similar return types.

### PR Checklist

- [ ] Related to designs:
- [x] Related to JIRA ticket: DISC-787
- [x] I have run this code to verify it works
- [ ] This PR includes unit tests for the code change
- [ ] This PR includes testing instructions tests for the code change
- [ ] The alpha package of this PR is passing end-to-end tests in all relevant Codecademy repositories

#### Testing Instructions

Make sure to delete cookies (for exp assignment) and view https://tengu.codecademy.com/catalog?PR_ENV=portal-app-pr-2426 as an anonymous user.  Can also add youself to the experiment for the logged in version https://app.optimizely.com/v2/projects/21056540090/experiments/22003060387/whitelist. Everything should just work, there is no visible change - just a refactor.

#### PR Links and Envs

https://tengu.codecademy.com/catalog?PR_ENV=portal-app-pr-2426

### CODE REVIEW

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.6.181-alpha.4ffe31.0<br/>@codecademy/gamut-labs@33.1.2-alpha.4ffe31.0<br/>@codecademy/styleguide@61.1.11-alpha.4ffe31.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://62fe7f4ad194af007e6cb7e5--gamut-preview.netlify.app>https://62fe7f4ad194af007e6cb7e5--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/62fe7f4ad194af007e6cb7e5>Deploy Logs</a></p>

File: packages/gamut-labs/src/GlobalHeader/index.tsx [hunk 1]
```diff
@@ -41,98 +41,96 @@ export type GlobalHeaderProps =
   | ProHeader
   | LoadingHeader;
 
-const getAppHeaderItems = (
-  props: GlobalHeaderProps
-): FormattedAppHeaderItems => {
+// Overloading getAppHeaderItems function to return different types based on mobile parameter
```
Comment (medhatnar) [inline packages/gamut-labs/src/GlobalHeader/index.tsx:44]: interesting. how is this working? You can have multiple typings for the same function?

Comment (j10sanders) [reply packages/gamut-labs/src/GlobalHeader/index.tsx:44]: Yeah, it's a new pattern for me as well, called function overloading. https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads

**Review by medhatnar — APPROVED** [review]
LGTM after fixing the failed job

File: packages/gamut-labs/src/GlobalHeader/index.tsx [hunk 2]
```diff
@@ -41,98 +41,96 @@ export type GlobalHeaderProps =
...
+            return anonLandingMobileHeaderItems(
+              hidePricing,
+              props.user,
+              renderBookmarks
+            );
+          }
           return anonLandingHeaderItems(
-            props.hidePricing,
+            hidePricing,
             props.user,
```
Comment (zippyzow) [inline packages/gamut-labs/src/GlobalHeader/index.tsx:73]: curious why you didn't destructure `user` as you did with `hidePricing` and `renderBookmarks`?

**Review by zippyzow — APPROVED** [review]
TIL about function overloads! LGTM and working as expected. Left a non-blocking question.

File: packages/gamut-labs/src/GlobalHeader/index.tsx [hunk 3]
```diff
@@ -41,98 +41,96 @@ export type GlobalHeaderProps =
...
+function getAppHeaderItems(
+  props: GlobalHeaderProps,
+  mobile: Boolean
+): FormattedAppHeaderItems | FormattedMobileAppHeaderItems {
+  const { renderBookmarks, hidePricing } = props;
   switch (props.type) {
     case 'anon':
       switch (props.variant) {
         case 'landing':
+          if (mobile) {
```
Comment (zippyzow) [inline packages/gamut-labs/src/GlobalHeader/index.tsx:64]: One more thought: What do you think about drying this up even further by utilizing this `mobile` param value by doing something like:

Creating one function per variant in GlobalHeaderVariants.tsx such as:
```
export const anonDefaultHeaderItems = (
  mobile: boolean,
  hidePricing?: boolean,
  user?: User,
  renderBookmarks?: () => ReactNode
): FormattedAppHeaderItems | FormattedMobileAppHeaderItems => {
  return mobile
    ? anonMobileHeaderItems(true, true, hidePricing, user, renderBookmarks)
    : anonHeaderItems(true, true, hidePricing, user, renderBookmarks);
};
```
and then calling it in this file like so...
```
switch (props.variant) {
        case 'landing':
          return anonDefaultHeaderItems(
            mobile,
            hidePricing,
            props.user,
            renderBookmarks
          );
```

File: packages/gamut-labs/src/GlobalHeader/index.tsx (hunk 2 above)
Comment (j10sanders) [reply packages/gamut-labs/src/GlobalHeader/index.tsx:73]: user isn't guaranteed to be in the object

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.6.181-alpha.c35a46.0<br/>@codecademy/gamut-labs@33.1.2-alpha.c35a46.0<br/>@codecademy/styleguide@61.1.11-alpha.c35a46.0<br/></pre>

File: packages/gamut-labs/src/GlobalHeader/index.tsx (hunk 3 above)
Comment (j10sanders) [reply packages/gamut-labs/src/GlobalHeader/index.tsx]: the renderBookmarks type is different between mobile and desktop, so it's going to get more complicated.  I added the ternary style to the function I already edited to get us close to that, but I think doing this full refactor is going to make it more confusing.  I like the suggestion though

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://63053c6ee47a2d0b28410807--gamut-preview.netlify.app>https://63053c6ee47a2d0b28410807--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/63053c6ee47a2d0b28410807>Deploy Logs</a></p>

**Pull request merged**